python/ray/dag/compiled_dag_node.py

```diff
@@ -2400,12 +2474,19 @@ def build_compiled_dag_from_ray_dag(
         overlap_gpu_communication,
     )
 
-    def _build_compiled_dag(node):
-        compiled_dag._add_node(node)
-        return node
-
     root = dag._find_root()
-    root.traverse_and_apply(_build_compiled_dag)
+    topo_queue = root.get_topo_queue()
+    node_to_p2p_send_node: Dict[DAGNode, _P2PSendNode] = dict()
+    for node in topo_queue:
+        if isinstance(node, _P2PNode):
+            raise ValueError(
+                "Please use type hints to specify NCCL transport instead of "
+                "adding _P2PSendNode or _P2PRecvNode to the DAG"
+            )
+        compiled_dag._add_nccl_p2p_recv_nodes(node, node_to_p2p_send_node)
```

Comment:
This looks a bit complicated right now and seems tricky to get all the bind indices, etc, correct. What if we add the NCCL p2p nodes when creating the initial DAG, instead of adding them after the DAG has already been created? That way we can append to the DAG instead of modifying in place.

Reply:
Do we expect users to add type hint right after a node is created instead of adding type hint later? Then when adding the type hint, we immediately create a NCCL send node.

Reply:
Options:
1. Create the NCCL nodes on the fly, while user is defining the DAG. Potential complexity: may not know if you're using NCCL yet?
2. Create the NCCL nodes while converting from the user's DAG to CompiledDAG.